dm/dm/master/openapi_controller.go

```diff
@@ -0,0 +1,288 @@
+
+package master
+
+import (
+	"context"
+
+	"github.com/pingcap/tiflow/dm/checker"
+	"github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/dm/pb"
+	"github.com/pingcap/tiflow/dm/openapi"
+	"github.com/pingcap/tiflow/dm/pkg/terror"
+)
+
+func (s *Server) getSourceStatusListFromWorker(ctx context.Context, sourceName string, specifiedSource bool) ([]openapi.SourceStatus, error) {
+	workerStatusList := s.getStatusFromWorkers(ctx, []string{sourceName}, "", specifiedSource)
+	sourceStatusList := make([]openapi.SourceStatus, len(workerStatusList))
+	for i, workerStatus := range workerStatusList {
+		if workerStatus == nil {
+			// this should not happen unless the rpc in the worker server has been modified
+			return nil, terror.ErrOpenAPICommonError.New("worker's query-status response is nil")
+		}
+		sourceStatus := openapi.SourceStatus{SourceName: sourceName, WorkerName: workerStatus.SourceStatus.Worker}
+		if !workerStatus.Result {
+			sourceStatus.ErrorMsg = &workerStatus.Msg
+		} else if relayStatus := workerStatus.SourceStatus.GetRelayStatus(); relayStatus != nil {
+			sourceStatus.RelayStatus = &openapi.RelayStatus{
+				MasterBinlog:       relayStatus.MasterBinlog,
+				MasterBinlogGtid:   relayStatus.MasterBinlogGtid,
+				RelayBinlogGtid:    relayStatus.RelayBinlogGtid,
+				RelayCatchUpMaster: relayStatus.RelayCatchUpMaster,
+				RelayDir:           relayStatus.RelaySubDir,
+				Stage:              relayStatus.Stage.String(),
+			}
+		}
+		sourceStatusList[i] = sourceStatus
+	}
+	return sourceStatusList, nil
+}
+
+// nolint:unparam
+func (s *Server) createSource(ctx context.Context, cfg *config.SourceConfig) error {
+	return s.scheduler.AddSourceConfig(cfg)
+}
+
+// nolint:unparam,unused
+func (s *Server) updateSource(ctx context.Context, cfg *config.SourceConfig) error {
+	// TODO(ehco) no caller now , will implement later
+	return nil
+}
+
+// nolint:unparam
+func (s *Server) deleteSource(ctx context.Context, sourceName string, force bool) error {
+	if force {
+		for _, taskName := range s.scheduler.GetTaskNameListBySourceName(sourceName) {
+			if err := s.scheduler.RemoveSubTasks(taskName, sourceName); err != nil {
+				return err
+			}
+		}
+	}
+	return s.scheduler.RemoveSourceCfg(sourceName)
+}
+
+// nolint:unparam,unused
+func (s *Server) getSource(ctx context.Context, sourceName string) (openapiSource openapi.Source, err error) {
+	sourceCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if sourceCfg == nil {
+		return openapiSource, terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	openapiSource = config.SourceCfgToOpenAPISource(sourceCfg)
+	return openapiSource, nil
+}
+
+func (s *Server) getSourceStatus(ctx context.Context, sourceName string) ([]openapi.SourceStatus, error) {
+	return s.getSourceStatusListFromWorker(ctx, sourceName, true)
+}
+
+// nolint:unparam
+func (s *Server) listSource(ctx context.Context, req interface{}) []openapi.Source {
```

Comment:
all usage of funcs in this file is not steable, parameters and return response will be enriched in the next pr